src/tally.cpp

```diff
@@ -420,34 +448,103 @@ std::string pyne::Tally::mcnp(int tally_index, std::string mcnp_version) {
 
   } else if (entity_type.find("Volume") != std::string::npos) {
     if (tally_type.find("Flux") != std::string::npos) {
-      output << form_mcnp_tally(tally_index, 4, particle_token, 
-                                entity_id, entity_size, normalization);
+      tally_id = 4;
     } else if (tally_type.find("Current") != std::string::npos) {
       // makes no sense in mcnp
+      return "";
+    }
+    output << form_mcnp_tally(tally_index, tally_id, particle_token, entity_id,
+                              entity_size, normalization);
+  } else if (entity_type.find("Mesh") != std::string::npos) {
+    output << "FMESH" << tally_index << "4:" << particle_token << "  ";
+    output << "GEOM=";
+    std::string indent_block = "           ";
+    std::stringstream sup_var;
+
+    if (entity_geometry.find("Cartesian") != std::string::npos) {
+      output << "XYZ ";
+    } else if (entity_geometry.find("Cylinder") != std::string::npos) {
+      output << "CYL ";
+      if (!is_zero(axl)) {
+        sup_var << indent_block << "AXL=" << mcnp_list_to_string(axl);
+      }
+      if (!is_zero(vec)) {
+        sup_var << "\n" << indent_block;
+        sup_var << "VEC=" << mcnp_list_to_string(vec);
+      }
+    }
+
+    output << " ORIGIN=" << mcnp_list_to_string(origin) << "\n";
+    std::string dir_name[3] = {"I", "J", "K"};
+    std::vector<double> meshes[3] = {i_meshs, j_meshs, k_meshs};
+    std::vector<int> bins[3] = {i_bins, j_bins, k_bins};
+
+    for (int j = 0; j < 3; j++) {
+      output << indent_block;
+      output << dir_name[j] << "MESH=" << mcnp_list_to_string(meshes[j]);
+      if (bins[j].size() > 0) {
+        output << " " << dir_name[j] << "INTS=" << mcnp_list_to_string(bins[j]);
+      }
+      output << "\n";
+    }
+    if (sup_var.str().size() > 0) {
+      output << sup_var.str() << "\n";
+    }
+    if (e_bounds.size() > 0) {
+      output << indent_block << "EMESH=";
+      output << mcnp_list_to_string(e_bounds);
+    }
+    output << "\n";
+    if (e_bins.size() > 0) {
+      output << indent_block << "EINTS=";
+      output << mcnp_list_to_string(e_bins);
+    }
+    if (out.size() > 0) {
+      output << "\n" << indent_block << "OUT=" << out;
     }
   } else {
-    std::cout << "tally/entity combination makes no sense for MCNP" << std::endl;
+    std::cout << "tally/entity combination makes no sense for MCNP"
+              << std::endl;
   }
-  
-
   // print sd card if area/volume specified
   return output.str();
-} 
+}
+template <typename T>
+bool pyne::Tally::is_zero(T vect) {
+  int size = sizeof(vect) / sizeof(vect[0]);
+  bool result = true;
+  for (int i = 0; i < size; i++) result &= (vect[i] == 0);
+  return result;
+}
+
+template<typename T>
+std::string pyne::Tally::mcnp_list_to_string(std::vector<T> vect){
+  std::stringstream out;
+  out << std::setiosflags(std::ios::fixed) << std::setprecision(6);
+  if (normalization > 1.0) 
+    out << std::scientific;
+  
+  for( auto elmt : vect)
```

Comment:
the auto keyword should be banned!

Reply:
Why??

Reply:
Replying to [this comment](https://github.com/pyne/pyne/pull/1153#issuecomment-515799377) - yeah I can see that at times. You can run into some weird implicit type conversion too. I do think it’s valuable in for loops like this for the decreased verbosity and also in those cases like you’ve cited for other relatively verbose types like `std::pair<T,V>`.

Reply:
I understand @makeclean point, I am working with some other codes with thousands of auto... this is a pain... really!

but I really like the usage of `auto` for loop, it simplifies the syntax (`int i .... foo[i]....` is a bit verbosy for me....) and allows to loop over an array in a python way which I like...
but I don't have a strong opinion... specially in this case...
I'll switch back on the conservative way... :)